### ISSUE TITLE
splitChunks with innerGraph does not add required dependencies into common chunk

### ISSUE THREAD

**Issue by Knagis** [webpack/webpack#11770]
# Bug report

**What is the current behavior?**

Two entries and common chunk (see the linked repo for more details):
* index1 - imports dep1 which imports dep2. actually uses both of them
* index2 - imports dep1 but only uses a function that does not depend on dep2
* commons - contains everything that is used by more than one entry

The issue is that dep2 is detected only as required only for index1 (since index2 does not actually use it). By itself that is kind of fine - when loading commons+index1, everything works, because dep2 is loaded from index1 chunk.

However when loading index2, fails with runtime error due to the fact that in commons chunk, dep1 still lists dep2 as dependency - but it is not included in either commons or index2.

**If the current behavior is a bug, please provide the steps to reproduce.**

https://github.com/Knagis/webpack-concat-issue/tree/commons-error

**What is the expected behavior?**

no runtime errors (either dep2 is included in commons chunk or it is loaded only when calling the function that uses it or various other combinations of moving/splitting the modules so that it works for both entries)

**Other relevant information:**
webpack version:  5.1.3
Node.js version: 12.13.1
Operating System: Win10
Additional tools:

**Comment by Knagis** [comment]
setting `optimization.innerGraph = false` prevents the problem - and gets both dep1 and dep2 added to the commons chunk. or adds dep2 to both index1 and index2 if I set the commons to explicitly only include dep1 (instead of all modules in more than one entry).

**Comment by code1x1** [comment]
i can reproduce it

https://github.com/Knagis/webpack-concat-issue/blob/c46dfd7cdfee8d301547a32417ca825c343e398e/dist/main2.js#L35

`__webpack_modules__` in main2.js only contains 51 and 113 but gets called with 891 
```
__webpack_modules__
{…}
​
51: function 51(__unused_webpack_module, __unused_webpack___webpack_exports__, __webpack_require__)​
113: function 113(__unused_webpack_module, __webpack_exports__, __webpack_require__)​
<prototype>: Object { … }

moduleId
891
__webpack_modules__[moduleId]
undefined
```

**Comment by sokra** [comment]
hmm, yes that's a tricky case... The import is actually unused in `index2`, but merging the two variants of `dep1` into a common variant (shared by both runtimes) generates an invalid `__webpack_require__(891)`.

We probably need to generate some code to conditionally import the dependencies depending on runtime, so these forcefully shared modules can be generated correctly.

There are multiple workaround possible:
* no `name` for `splitChunks` to avoid forcing them to merge
* `optimization.usedExports: "global"` to disable runtime-dependent exports
* `optimization.innerGraph: false` to disable import usage tracking

**Comment by doberkofler** [comment]
I'm hitting the same problem. Using `optimization.sideEffects: false` seems to work around the issue for me.

**Comment by sokra** [comment]
There is a bugfix here: https://github.com/webpack/webpack/pull/11802
but I'm unsure if this has more undiscovered side effects...

Maybe you can test the PR in your real repos, that would give me more certainty that it works fine?

You can just reference a branch in your package.json `"webpack": "webpack/webpack#bugfix/11770"`

**Comment by Knagis** [comment]
will do. hopefully during weekend, if not, on Monday.

**Comment by doberkofler** [comment]
@sokra I just tested the the `bugfix/11770` branch  and unfortunately it shows about 100 errors like the following.
Is there anything I can do to help?

```
ERROR in ./src/framework/SplitViewModule/configuration.ts
str.replace is not a function
 @ ./src/framework/SplitViewModule/index.ts 9:0-56 104:22-43
 @ ./src/modules/las_mod_siheader/index.ts 16:0-79 45:36-51 106:27-41 140:27-42 149:27-41 154:32-47 166:93-108 180:27-42 206:17-32 213:17-32 229:17-31 233:17-32 538:27-42 538:57-71

ERROR in ./src/framework/SplitViewModule/info.ts
str.replace is not a function
 @ ./src/framework/SplitViewModule/index.ts 16:0-38 211:15-27
 @ ./src/modules/las_mod_siheader/index.ts 16:0-79 45:36-51 106:27-41 140:27-42 149:27-41 154:32-47 166:93-108 180:27-42 206:17-32 213:17-32 229:17-31 233:17-32 538:27-42 538:57-71

ERROR in ./src/framework/SplitViewModule/table.ts
str.replace is not a function
 @ ./src/framework/SplitViewModule/index.ts 15:0-152 134:15-30 164:32-43 198:19-31 422:47-64 445:16-33 470:16-33 505:20-37 506:54-67 546:46-59 639:45-62 640:39-52 663:37-50 681:43-60
 @ ./src/modules/las_mod_siheader/index.ts 16:0-79 45:36-51 106:27-41 140:27-42 149:27-41 154:32-47 166:93-108 180:27-42 206:17-32 213:17-32 229:17-31 233:17-32 538:27-42 538:57-71

ERROR in ./src/framework/UniversalTable/columnsMenu.ts
str.replace is not a function
 @ ./src/framework/UniversalTable/index.ts 10:0-50 189:41-58
 @ ./src/framework/SplitViewModule/index.ts
 @ ./src/modules/las_mod_siheader/index.ts 16:0-79 45:36-51 106:27-41 140:27-42 149:27-41 154:32-47 166:93-108 180:27-42 206:17-32 213:17-32 229:17-31 233:17-32 538:27-42 538:57-71
```

**Comment by sokra** [comment]
Could you add `stats.errorDetails: true` or `stats.errorStack: true` to get a stack trace?

**Comment by doberkofler** [comment]
@sokra Does this help
```
ERROR in ./src/framework/SplitViewModule/configuration.ts
str.replace is not a function
TypeError: str.replace is not a function
    at quoteMeta (/Users/doberkofler/MyDev/ljs_app/trunk/periscope/node_modules/webpack/lib/util/compileBooleanMatcher.js:9:13)
    at Function.from (<anonymous>)
    at itemsToRegexp (/Users/doberkofler/MyDev/ljs_app/trunk/periscope/node_modules/webpack/lib/util/compileBooleanMatcher.js:195:49)
    at Function.compileBooleanMatcherFromLists [as fromLists] (/Users/doberkofler/MyDev/ljs_app/trunk/periscope/node_modules/webpack/lib/util/compileBooleanMatcher.js:41:25)
    at RuntimeTemplate.runtimeConditionExpression (/Users/doberkofler/MyDev/ljs_app/trunk/periscope/node_modules/webpack/lib/RuntimeTemplate.js:592:32)
    at RuntimeTemplate.importStatement (/Users/doberkofler/MyDev/ljs_app/trunk/periscope/node_modules/webpack/lib/RuntimeTemplate.js:625:28)
    at HarmonyImportSpecifierDependency.getImportStatement (/Users/doberkofler/MyDev/ljs_app/trunk/periscope/node_modules/webpack/lib/dependencies/HarmonyImportDependency.js:90:26)
    at HarmonyImportSpecifierDependencyTemplate.apply (/Users/doberkofler/MyDev/ljs_app/trunk/periscope/node_modules/webpack/lib/dependencies/HarmonyImportDependency.js:290:31)
    at HarmonyImportSpecifierDependencyTemplate.apply (/Users/doberkofler/MyDev/ljs_app/trunk/periscope/node_modules/webpack/lib/dependencies/HarmonyImportSpecifierDependency.js:298:10)
    at JavascriptGenerator.sourceDependency (/Users/doberkofler/MyDev/ljs_app/trunk/periscope/node_modules/webpack/lib/javascript/JavascriptGenerator.js:203:12)
    at JavascriptGenerator.sourceModule (/Users/doberkofler/MyDev/ljs_app/trunk/periscope/node_modules/webpack/lib/javascript/JavascriptGenerator.js:112:9)
    at JavascriptGenerator.generate (/Users/doberkofler/MyDev/ljs_app/trunk/periscope/node_modules/webpack/lib/javascript/JavascriptGenerator.js:98:8)
    at NormalModule.codeGeneration (/Users/doberkofler/MyDev/ljs_app/trunk/periscope/node_modules/webpack/lib/NormalModule.js:924:22)
    at /Users/doberkofler/MyDev/ljs_app/trunk/periscope/node_modules/webpack/lib/Compilation.js:2147:24
    at /Users/doberkofler/MyDev/ljs_app/trunk/periscope/node_modules/webpack/lib/CacheFacade.js:51:43
    at /Users/doberkofler/MyDev/ljs_app/trunk/periscope/node_modules/webpack/lib/Cache.js:93:5
 @ ./src/framework/SplitViewModule/index.ts 9:0-56 104:22-43
 @ ./src/modules/las_mod_user/index.ts 21:0-79 40:32-47 134:21-36 135:21-36 138:21-35 141:21-36 187:96-110 188:27-41 283:17-32 293:17-32 323:17-31 327:17-32
```

**Comment by sokra** [comment]
Yep, thanks that helped. I updated the PR with a fix for that and a few other fixes I discovered. If you have some time it would be cool to test the PR again. If it works for you, I would publish a new webpack version for this.

**Comment by doberkofler** [comment]
@sokra I'm working on it ...

**Comment by doberkofler** [comment]
I have now tested the update PR and the `TypeError: str.replace is not a function` have disappeared but unfortunately the generated bundle does still not run. Interestingly the bundle did work before when using the `optimization.sideEffects: false` option but now also fails with this option.

The error stack looks like this:
![image](https://user-images.githubusercontent.com/1009585/97203742-88687a80-17b5-11eb-8b51-b0bfa3e05d0f.png)

The actual code where the error is reported looks like this:
![image](https://user-images.githubusercontent.com/1009585/97203926-c2d21780-17b5-11eb-982d-38c535a466b1.png)

Is there anything I can do to help?

**Comment by Knagis** [comment]
In my case the error I used to get due to the common chunk seems to be fixed.

**Comment by Knagis** [comment]
@doberkofler did you try using `innerGraph: false` as a workaround? That is how I worked around the issue in my repo and it seems to be better than completely turning off sideEffects - at least compared to webpack 4 there wouldn't be a loss of optimizations.

**Comment by doberkofler** [comment]
I have also tried `innerGraph: false` but it still generated a bundle does not run. Maybe I'm hitting a different bug?

**Comment by evilebottnawi** [comment]
@doberkofler yep, maybe... Can you provide example of code or reproducible repo or maybe you can simplify to provide simple example

**Comment by doberkofler** [comment]
@evilebottnawi As I wrote something must have changed in this PR because now the `optimization.sideEffects: false` workaround is no longer working as well. I would love to provide a simple example but unfortunately most of my problem only appear in one of my very large production environments. Do the errors I get indicate anything that might help you to identify what goes wrong or help me to understand how to create a simple example ?

**Comment by sokra** [comment]
Could you disable source map and look at the `extends UiElement` code (the generated code)?

**Comment by doberkofler** [comment]
The line itself looks like this
```js
class UiElementFile extends (/* unused pure expression or super */ 7324 == __webpack_require__.j && (UiElement_UiElement)) {
```

**Comment by sokra** [comment]
Yep that's what I through. Ill fix it...

**Comment by doberkofler** [comment]
... fortunately I don't need to understand ;-)

**Comment by sokra** [comment]
@doberkofler ok fixed. Could you try once again?

**Comment by doberkofler** [comment]
@sokra I had to reinstall all modules after completely removing my node_modules directory and the package-lock.json file before seeing any difference but unfortunately I now get a different error.

![image](https://user-images.githubusercontent.com/1009585/97265558-247e9a00-1827-11eb-91e7-ecc3e80c11ce.png)

in this line
```js
/******/ 		__webpack_modules__[moduleId].call(module.exports, module, module.exports, __webpack_require__);
```

Is there anything else I can do to help ?

**Comment by sokra** [comment]
Ok. So the `extends UiElement` is fixed and this is another one...

You could add `optimization: { minimize: false, moduleIds: "named" }` to the config to get better stack traces. Interesting would be the location of the bundle where the `__webpack_require__` is called. (here: `at Object.47916`, but it's different with the different settings)

**Comment by sokra** [comment]
Maybe I add a little explanation why this issue is so complex:

Since webpack 5 webpack optimizes used exports per entrypoint. So if entrypoint A uses export X of module M and entrypoint B uses export Y of module M, then webpack will optimize the graph for each entrypoint separately. This means if export X of module M references module N, then module N will only be part of the graph of entrypoint A.

There will be two variants of module M: let's call them Mx and My. Mx only includes export X and references module N. My only includes export Y and doesn't reference module N.

This all works fine util Mx and My are merged into a single chunk (by a splitChunks configuration). Merging them makes sense in some cases, e. g. to reduce duplication and improve long term caching.

Currently webpack merges the variants into a combined variant Mxy which includes exports X and Y and references module N. The problem occurs when module N is not merged into the combined chunk, but kept somewhere else. In this case module N is only available in graph A and not in graph B. Executing Mxy in graph B fails, because module N is not available. This causes the `Cannot read property 'call' of undefined` error.

My PR adds the ability to generate code that executes differently depending on which is the current graph (called 'runtime' in webpack). So basically it generates code like `if (graph == A) reference(N)`. This allows that module Mxy executes like Mx in graph A and like My in graph B. Such code is only generated for some pieces of code:

* `import { ... } from "side-effect-free-module"`: The module might be referenced or not.
* `const x = expression;`: if `expression` is pure and `x` is only used by exports that are used or not.
  * We need to skip evaluation of expression as it might contain imported references.
* `class X extends expression`: if `expression is pure and `X` is only used exports that are used or not.
  * Similar to above
  * It's a bit tricky as `expression` must only replaced with `null` since only `null` is a valid super class.

**Comment by doberkofler** [comment]
> Ok. So the `extends UiElement` is fixed and this is another one...

I guess so

> You could add `optimization: { minimize: false, moduleIds: "named" }` to the config to get better stack traces. Interesting would be the location of the bundle where the `__webpack_require__` is called. (here: `at Object.47916`, but it's different with the different settings)

`minimize: false` was already set and i now also set `moduleIds: "named"` with the following results

**this is the stack**
![image](https://user-images.githubusercontent.com/1009585/97270173-04eb6f80-182f-11eb-88a0-4695fe8c5371.png)

**at __webpack_require__ (las_mod_dashboard.min.js?_=20.10.0:25879)**
`/******/ 		__webpack_modules__[moduleId].call(module.exports, module, module.exports, __webpack_require__);`

**at Object../src/framework/ods/date.ts (vendors.min.js?_=20.10.0:195129)**
`var add = __webpack_require__("./node_modules/date-fns/esm/add/index.js");`

**at __webpack_require__ (las_mod_dashboard.min.js?_=20.10.0:25879)**
`/******/ 		__webpack_modules__[moduleId].call(module.exports, module, module.exports, __webpack_require__);`

**at Object../src/framework/uielement/private/UiElementInputDate.ts (vendors.min.js?_=20.10.0:204911)**
`/* harmony import */ var Source_ODS_date__WEBPACK_IMPORTED_MODULE_0__ = __webpack_require__("./src/framework/ods/date.ts");`

...

**Comment by sokra** [comment]
Ok I think this is related to module concatentation, which doesn't seem to handle conditional imports correctly.

I guess if you add `optimization.concatenateModule: false` the error disappears?

**Comment by doberkofler** [comment]
I guess it was the `concatenateModules` option and yes, by disabling module concatenation the error disappears.

**Comment by doberkofler** [comment]
> Maybe I add a little explanation why this issue is so complex:
> 
> Since webpack 5 webpack optimizes used exports per entrypoint. So if entrypoint A uses export X of module M and entrypoint B uses export Y of module M, then webpack will optimize the graph for each entrypoint separately. This means if export X of module M references module N, then module N will only be part of the graph of entrypoint A.
> 
> There will be two variants of module M: let's call them Mx and My. Mx only includes export X and references module N. My only includes export Y and doesn't reference module N.
> 
> This all works fine util Mx and My are merged into a single chunk (by a splitChunks configuration). Merging them makes sense in some cases, e. g. to reduce duplication and improve long term caching.
> 
> Currently webpack merges the variants into a combined variant Mxy which includes exports X and Y and references module N. The problem occurs when module N is not merged into the combined chunk, but kept somewhere else. In this case module N is only available in graph A and not in graph B. Executing Mxy in graph B fails, because module N is not available. This causes the `Cannot read property 'call' of undefined` error.
> 
> My PR adds the ability to generate code that executes differently depending on which is the current graph (called 'runtime' in webpack). So basically it generates code like `if (graph == A) reference(N)`. This allows that module Mxy executes like Mx in graph A and like My in graph B. Such code is only generated for some pieces of code:
> 
> * `import { ... } from "side-effect-free-module"`: The module might be referenced or not.
> * `const x = expression;`: if `expression` is pure and `x` is only used by exports that are used or not.
>   
>   * We need to skip evaluation of expression as it might contain imported references.
> * `class X extends expression`: if `expression is pure and `X` is only used exports that are used or not.
>   
>   * Similar to above
>   * It's a bit tricky as `expression` must only replaced with `null` since only `null` is a valid super class.

Thank you for the explanation that I most likely did not fully understand ;-)
It is quite stunning how complex a seemingly simple task like bundling can be and unfortunately most people are not aware of the complexity and just take the results as granted.
Thank you for your great work! (I'm happy to at least contribute a little)

**Comment by sokra** [comment]
> I guess it was the `concatenateModules` option and yes, by disabling module concatenation the error disappears.

Great. I work on fixing that thing too...

**Comment by sokra** [comment]
> It is quite stunning how complex a seemingly simple task like bundling can be and unfortunately most people are not aware of the complexity and just take the results as granted.

Most complexity doesn't come from a single feature. Most features along are pretty straight-forward. Most complexity comes from combining multiple features so that they play well together. This issue is a good example. It only occurs as a combination of:

* Code Splitting and especially custom chunk merging via `splitChunks`
* used exports analysis per runtime
* side-effect-free modules
* analysis of dependencies between top-level-symbols in modules
* and now also: concatenation of modules

**Comment by sokra** [comment]
> Great. I work on fixing that thing too...

Ok, this was a big one... Should be fixed now. Maybe you can try it again? Sorry for bothering you so often...

**Comment by sokra** [comment]
> I had to reinstall all modules after completely removing my node_modules directory and the package-lock.json file before seeing any difference but unfortunately I now get a different error.

yeah npm and yarn doesn't really like if the `version` in package.json doesn't change.

**Comment by doberkofler** [comment]
>Ok, this was a big one... Should be fixed now. Maybe you can try it again? Sorry for bothering you so often...

That's the least I can do without having provided a simple example.
I guess something it comes handy to have a huge and complicated production environment ;-)
I switched back to the original configuration and it works like a charm now.
Congratulations!

**Comment by doberkofler** [comment]
> yeah npm and yarn doesn't really like if the `version` in package.json doesn't change.

The only way for me: removing the `node_modules` directory and the `package-lock.json` file.

**Comment by sokra** [comment]
> I switched back to the original configuration and it works like a charm now.

great. released as [5.3.0](https://github.com/webpack/webpack/releases/tag/v5.3.0)

**Comment by sokra** [comment]
Thanks for all your testing and comments

**Comment by doberkofler** [comment]
You are welcome

**Comment by Knagis** [comment]
I might have spoken too soon... With 5.3.0 still getting these same errors. Will now try to pinpoint what is the cause.

**Comment by doberkofler** [comment]
@Knagis Just an idea but I had to manually remove the `node_modules` directory and the `package-lock.json` file to successfully update to the new npm version.

**Comment by Knagis** [comment]
@doberkofler I had the error from fresh CI build, so nothing cached is to be blamed.

@sokra The error I get now is due to:
```
// EXTERNAL MODULE: ./packages/path/SomeModule.ts
var SomeModule = __webpack_require__(null);
```

obviously moduleId `null` does not work... This module is one that is extracted into a common chunk as part of concatenated module with ID=784404 - an ID that is not referenced anywhere at all.

Should I report another issue or reopen this?

**Comment by evilebottnawi** [comment]
@Knagis let's open a new issue, I think this is a related problem, but not the same
